src/libraries/Microsoft.Extensions.FileSystemGlobbing/src/Internal/PatternContexts/CompositePatternContext.cs

```diff
@@ -0,0 +1,27 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using Microsoft.Extensions.FileSystemGlobbing.Abstractions;
+
+namespace Microsoft.Extensions.FileSystemGlobbing.Internal.PatternContexts
+{
+    internal abstract class CompositePatternContext : IPatternContext
+    {
+        public abstract void Declare(Action<IPathSegment, bool> onDeclare);
+        public abstract void PopDirectory();
+        public abstract void PushDirectory(DirectoryInfoBase directory);
+
+        protected internal abstract PatternTestResult MatchPatternContexts<TFileInfoBase>(
+            TFileInfoBase fileInfo,
+            Func<IPatternContext, TFileInfoBase, PatternTestResult> test);
```

Comment:
Not need to be generic.
```suggestion
        protected internal abstract PatternTestResult MatchPatternContexts(
            FileSystemInfoBase fileSystemInfo,
            Func<IPatternContext, PatternTestResult> test);
```

Reply:
The type argument's naming is confusing since we don't actually care that it's a file info. This is instead using a `TState` pattern, so the implementation will just call the `test` with their `IPatternContext` and pass in this opaque state for the ride. It could also just be:

```cs
protected internal abstract PatternTestResult MatchPatternContexts(Func<IPatternContext, PatternTestResult> test);
```

which would force us to have a closure. I don't know if that matters so much here, but this method is from the existing code and I decided to just leave it as is.